ops/framework.py

```diff
@@ -88,7 +88,48 @@ def from_path(cls, path):
         return handle
 
 
-class EventBase:
+class UnassignedEvent(type):
+    """Metaclass for events which have not yet been assigned to an emitter type.
+
+    Without being assigned and bound to an emitter, unassigned events cannot be emitted.
+    """
+
+    def __set_name__(event_type, emitter_type, event_kind):
+        # Permanently assign the event for this emitter type.
+        setattr(emitter_type, event_kind, AssignedEvent(event_type, event_kind, emitter_type))
+
+
+class AssignedEvent:
+    """An event which has been assigned to an emitter type with a name.
+
+    The assigned event is still not bound to an emitter instance, and thus cannot be emitted.
+    """
+
+    def __repr__(self):
+        return (f'<AssignedEvent {self.event_type.__name__} assigned as '
+                f'{self.event_kind} to {self.emitter_type} '
+                f'at {hex(id(self))}')
+
+    def __init__(self, event_type, event_kind, emitter_type):
+        self.event_type = event_type
+        self.event_kind = event_kind
+        self.emitter_type = emitter_type
+
+    def __get__(self, emitter, emitter_type):
+        if emitter:
+            # Permanently bind the event for this emitter.
+            bound_event = BoundEvent(emitter, self.event_type, self.event_kind)
+            setattr(emitter, self.event_kind, bound_event)
+            if emitter.framework:
+                # Register the event with the framework now that it is bound. The framework might not
+                # be bound if the event is accessed as a class attribute for introspection purposes.
+                emitter.framework.register_type(self.event_type, emitter, self.event_kind)
```

Comment:
Per comment above, lots of logic changing or moving around which feels unrelated to the PR purpose.